### ISSUE TITLE
[CUJ] git in notebooks

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#2889]
A common feature with notebooks is checking out and pushing code to git.

This is currently a painful manual process

   * Users need to manually clone the repos they want every time they start a notebook
   * Users need to manually upload git ssh keys or other credentials in order to be able to push to various repos

A CUJ might be a good place to start; 
    * e.g. in the notebook manager UI how could I specify repositories to clone on startup (see #2888) as well as Git credentials to inject into the pod?

Some thoughts:

* We are increasingly making use of Admission hooks to inject things like secrets
* We could use admission hooks to inject git config (e.g. secrets) and repos to clone (see #2888)
* Admission hooks can be attached to pods by attaching specific labels
* Could the notebook manager UI automatically detect which webhooks are available in the cluster and surface them in a meaningful way to users?
* For example in the notebook UI I see a list of options
 

- [ ] Attach repository github.com/kubeflow/examples
- [ ] Attach GCP credentials for account ..
- [ ] ...

Ideally the web UI dynamically detects and services admission hooks; e.g. for each admission hook we publish meta information such as a human readable description of what it does and the labels used to apply that hook. 

/cc @zabbasi @kimwnasptd @vkoukis

**Comment by jlewi** [comment]
Should we punt this to 0.7?

**Comment by geooff** [comment]
Any updates on the feature outlined above? My team is looking to deploy Kubeflow and are wondering if there are any established best practices regarding VC.

**Comment by jbottum** [comment]
@vkoukis could we consider this request for 1.0?

**Comment by karlschriek** [comment]
@jlewi , from what I can see [here](https://github.com/orgs/kubeflow/projects/25?card_filter_query=git), this feature has been included in the v1.0 release. However, I can't anything that describes how it works.

I am currently advising a team that are very interested in this feature (so much so that fluid git integration for notebooks is a make or break on whether they use kubeflow at all). Could you give me some direction on how this can be implemented?

**Comment by jlewi** [comment]
@karlschriek I don't think we have a working CUJ today. So it probably have to stitch things together.

Here are [instructions](https://github.com/jlewi/kubeflow-dev/blob/master/notes/git_on_kubeflow.md) for how I use git in notebooks today.

Its pretty manual today. 

Our [PodDefaults](https://github.com/kubeflow/kubeflow/tree/master/components/admission-webhook) admission hook could be used to inject a secret containing the SSH key but users would still have to run commands in a terminal to configure ssh to use it.

[PodDefaults](https://github.com/kubeflow/kubeflow/blob/master/components/admission-webhook/pkg/apis/settings/v1alpha1/poddefault_types.go) doesn't support init-containers today.

**Comment by jlewi** [comment]
Here are a couple of friction points with git I'm currently encountering.

* I need to start the ssh-agent every time I launch a new terminal.

  * This [recipe](https://stackoverflow.com/questions/18880024/start-ssh-agent-on-login) might help

* The permissions on my .ssh directory get reset (when the container restarts?)

**Comment by jlewi** [comment]
/lifecycle frozen

@kimwnasptd @mameshini @karlschriek This is the long standing issue for git in notebooks.
